src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/SignatureHelper.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+
+namespace System.Reflection.Emit.Experiment
+{
+    // This is prototype code, to generate simple signatures.
+    // For more complex signatures, will port System.Reflection.Emit's SignatureHelper.
+    internal static class MetadataSignatureHelper
+    {
+        internal static BlobBuilder FieldSignatureEncoder(Type fieldType)
+        {
+            var fieldSignature = new BlobBuilder();
+
+            MapReflectionTypeToSignatureType(new BlobEncoder(fieldSignature).FieldSignature(), fieldType);
+
+            return fieldSignature;
+        }
+
+        internal static BlobBuilder MethodSignatureEncoder(ParameterInfo[]? parameters, ParameterInfo? returnType, bool isInstance)
+        {
+            Type[]? typeParameters = null;
+            Type? typeReturn = null;
+
+            if (parameters != null)
+            {
+                typeParameters = Array.ConvertAll(parameters, parameter => parameter.ParameterType);
+            }
+
+            if (returnType != null)
+            {
+                typeReturn = returnType.ParameterType;
+            }
+
+            return MethodSignatureEncoder(typeParameters, typeReturn, isInstance);
+        }
+        internal static BlobBuilder MethodSignatureEncoder(Type[]? parameters, Type? returnType, bool isInstance)
+        {
+            // Encoding return type and parameters.
+            var methodSignature = new BlobBuilder();
+
+            ParametersEncoder _parEncoder;
+            ReturnTypeEncoder _retEncoder;
+
+            new BlobEncoder(methodSignature).
+                MethodSignature(isInstanceMethod: isInstance).
+                Parameters((parameters == null) ? 0 : parameters.Length, out _retEncoder, out _parEncoder);
+
+            if (returnType != null && returnType != typeof(void))
```

Comment:
Ths cannot use `typeof(void)`. It won't work for foreign type system.

Reply:
(I am not commenting on all uses of `typeof`.)

Reply:
> This cannot use typeof(void). It won't work for foreign type system.

Then I guess it should check the Type.FullName, same as in switch case below